### PR TITLE
Feature/improve cinterface

### PR DESCRIPTION
**Pull request by lukaspj**
# CInterface improvements
 * Expand EngineAPI type definitions (replaces #2241)
 * Fix name-clashing in script-exposed classes (these have not been an issue in TorqueScript because it doesn't fail, it just picks the first definition it finds).
 * Update the EngineAPI, a little bit of simplification over the changes made in #2248 
 * Improve the integration of the CInterface into existing systems
 * Expose SimPersistID to the EngineAPI
 * Improve EngineAPI export, less guess-work and compiler-dependency

# Breaking changes
Name-clash fixes means that some members have been renamed, possibly breaking any scripts using these members.
**AFX Members:**
 * _Method_ `afxPhraseEffectData::addEffect` -> pushEffect
 * _Method_ `afxMagicSpellData::addCastingEffect` -> pushCastingEffect
 * _Method_ `afxMagicSpellData::addLaunchEffect` -> pushLaunchEffect
 * _Method_ `afxMagicSpellData::addDeliveryEffect` -> pushDeliveryEffect
 * _Method_ `afxMagicSpellData::addImpactEffect` -> pushImpactEffect
 * _Method_ `afxMagicSpellData::addLingerEffect` -> pushLingerEffect
 * _Method_ `afxEffectronData::addEffect` -> pushEffect
 * _Method_ `afxEffectGroupData::addEffect` -> pushEffect

**Other Members:**
 * _Field_ `GuiSpeedometerHud::mColor` -> mNeedleColor (Maybe we could just use the parent class' mColor here instead?)
 * _Field_ `PostEffect::isEnabled` -> enabled
 * _Method_ `GuiFilterCtrl::identity` -> resetFiltering
 * _Field_ `SFXDescription::reverbModTime` -> reverbModDepth (dunno if it pointed to time or depth before, now there is one for each)
 * _Field_ `GuiScriptNotifyCtrl::onChildAdded` -> notifyOnChildAdded
 * _Field_ `GuiScriptNotifyCtrl::onChildRemoved` -> notifyOnChildRemoved
 * _Field_ `GuiScriptNotifyCtrl::onChildResized` -> notifyOnChildResized
 * _Field_ `GuiScriptNotifyCtrl::onParentResized` -> notifyOnParentResized
 * _Field_ `GuiScriptNotifyCtrl::onResize` -> notifyOnResize
 * _Field_ `GuiScriptNotifyCtrl::onLoseFirstResponder` -> notifyOnLoseFirstResponder
 * _Field_ `GuiScriptNotifyCtrl::onGainFirstResponder` -> notifyOnGainFirstResponder

## Other (possibly breaking) changes
I also changed the default value of `UndoManager::pushCompound` from "\"\"" to "". Would like a review of whether this is valid or a misunderstanding on my part.

Edit:
Some elaboration for why the name clashing is an issue.
The issue with the name clashing is that you have two members on the same object with the same name.
I.e. a class Foo with a method Bar and a property Bar.
If you doo f = new Foo();
f.Bar
Do you get the method or the property?

TorqueScript allows this, because it's TorqueScript, but other languages do not allow this, and it's generally a bad practice.

### CODE REVIEW

**Comment by dottools** [comment]
Best pull request comment ever. Nice and clear as to what's being changed.

**Comment by lukaspj** [comment]
@dottools thanks! I'm continually seeking to improve my Gittiquette :) 
In this one, I was extra aware of writing a good description, since it introduces a few breaking changes.

**Comment by lukaspj** [comment]
I have begun writing some documentation on my T3DSharp project, which can be used for testing the C-Interface:
https://github.com/lukaspj/T3DSharp/

If anyone wants to verify this PR, they could try going through the setup steps there and see if it makes sense.

**Comment by lukaspj** [comment]
Adding the "help wanted" label because having as many people review this as possible would be great :)

**Comment by lukaspj** [comment]
I've finally created a sample repo:
https://github.com/lukaspj/T3DSharp-Sample

It has a simple Coin Collection Module implementation in C#

File: Engine/source/afx/afxSelectron.cpp [hunk 1]
```diff
@@ -35,7 +35,7 @@
 //~~~~~~~~~~~~~~~~~~~~//~~~~~~~~~~~~~~~~~~~~//~~~~~~~~~~~~~~~~~~~~//~~~~~~~~~~~~~~~~~~~~~//
 // afxSelectronData::ewValidator
 //
-// When an effect is added using "addEffect", this validator intercepts the value
```
Comment (lukaspj) [inline Engine/source/afx/afxSelectron.cpp]: This line should be reverted

**Comment by lukaspj** [comment]
I rebased on top of newest version of the preview4_0 branch, and added @Azaezel 's commits to this one

**Comment by lukaspj** [comment]
I'll do a quick compile-and-run test of this. Don't merge yet

**Comment by lukaspj** [comment]
Build-and-run test is OK

**Pull request merged**